baseapp/abci.go

```diff
@@ -52,6 +52,8 @@ func (app *BaseApp) InitChain(req abci.RequestInitChain) (res abci.ResponseInitC
 		app.StoreConsensusParams(app.deliverState.ctx, req.ConsensusParams)
 	}
 
+	app.grpcQueryRouter.dryRunMethodHandlers(app.checkState.ctx)
```

Comment:
Status update: I added a `returnType map[string]reflect.Type` which maps each FQ method name to its return type. This map get populated when calling `dryRunMethodHandlers`, but i'm not sure from where to call it.

(Note: to dry run the method handlers, we need a `context.Context`, and I'm passing a `sdk.WrapSDKContext(ctx)` right now)

I tried:
- populating `returnTypes` immediately on `RegisterService` (so that we wouldn't even need a separate `dryRunMethodHandlers` function) using a dummy `sdk.Context`. But it panicked on some nil stores.
- calling `dryRunMethodHandlers` in InitChain, panicking for unmarshalling empty bytes request.

I'm thinking of a way to avoid panicking on all grpc query handlers. Maybe there's a way to do it with `context.CancelFunc` instead of `sdk.WrapSDKContext(someSdkCtx)`? Some other ideas?

Reply:
You can pre-run these in `RegisterService`. I'll take a quick look

Reply:
Actually looks pretty nasty... I remember we had this issue before.

Reply:
Actually if this is all the same process we can just capture the return type the first time a real query runs. I know that's super hacky but it should work unless someone is running grpc stand-alone. 

A more correct way would be - read grpc schema from message descriptors, then look up the return type name, then look up the actual go type from the proto type registry. 

What do you think?

Reply:
The grpc reflection way also seems more correct to me, however:

> then look up the return type name

hmm, I can't seem to find where to find the mapping of fq method name to return type (in grpc-go code)

> Actually if this is all the same process we can just capture the return type the first time a real query runs. I know that's super hacky but it should work unless someone is running grpc stand-alone.

For now I pushed this version

Reply:
Get the file descriptor with `proto.FileDescriptor` and to type with `proto.MessageType`. Does that help? This involves actually parsing the file descriptor...

Reply:
> This involves actually parsing a FileDescriptor

Ah right. I would need to define some heuristics to guess the file name from the fq method name, right? Will try that, but seems hacky too...

Reply:
I think it'll be hacky, we have e.g. `cosmos/base/reflection/v1beta1/reflection.proto`, and there's no way we can guess it from the fq service name (e.g. `cosmos.base.reflection.v1beta1.ReflectionService`), unless introducing a lot of ifs.

Maybe let's just stay with capturing the return type on 1st call?

Reply:
If we have access to a reflection grpc service then we can actually just use that to get the descriptors. But honestly hopefully this is just a short term fix and we deal with the concurrency issues soon. I would hate for the whole abci pipeline to be forever single threaded.

Reply:
> If we have access to a reflection grpc service then we can actually just use that to get the descriptors.

We'd need a client for that, it seems overkill. i also remember @clevinson running into issues with reflection because we're using gogoproto.

anyways, the current approach should be good enough for now.